Change the order of how nodes are attached to the client

gilberto-torrezan opened the pull request:
Now the index 0 considers the first server-side mapped node after all the client-side-only nodes.

This basically means that from now on all server-side nodes will be added after the preexisting client-side nodes, and not before as it previously did.

denis-anisimov commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kqg4R6v9JEIqUjIMz:-L8kqg4R6v9JEIqUjIN-:b-duy1mw) (no related file):*
There is no GWT unit for the situation when mapped element has only client side children (no any single mapped to the StateNode). And a new `StateNode` is added as a child.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1029 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kjmbABILpUlUe19Bn:-L8kjmbABILpUlUe19Bo:b967ivx) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1029)):*
> c

I understand that this is just an index but `c`  is very confusing. 
Either `j` or better: extract nested cycle to a separate method.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1033 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kp2Jl2I9apen_o_XV:-L8kp2Jl2I9apen_o_XW:b53ikki) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1033)):*
> if (object instanceof StateNode) {

That should always happen.
Either no check at all or assertion instead.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1035 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kpD-V7kbfD0NJhVMz:-L8kpD-V7kbfD0NJhVN-:b-9s5zzi) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1035)):*
>  if (clientNode.equals(stateNode.getDomNode())) {

Do you need double navigation?

`((StateNode)mappedNodeChildren.get(0)).getNode()` will give you the very first DOM node which is mapped as a `StateNode`.

And this should return the last DOM `Node` if  there is no children at all I think.

---

*[flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/template/ChildOrderView.java, line 64 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kr1xy2QMGWDMAK1sf:-L8kr1xy2QMGWDMAK1sg:b-7cxem1) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/template/ChildOrderView.java#L64)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>
>         addChildToContainer1.addClickListener(event -> {
>             Div div = new Div();
>             div.setText("Server child "
>                     + (containerWithElement.getElement().getChildCount() + 1));
>             containerWithElement.add(div);
>         });
</details>

As it shown in practice it's always good to have a test when action is done immediately in CTOR instead of as a click (or any other) listener.
It usually covers a listener case as well.
So please make at least one action immediately without a listener.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3773)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kqg4R6v9JEIqUjIMz:-L8lg8DI3iOneEwdViic:b-s1sgpi) (no related file):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

There is no GWT unit for the situation when mapped element has only client side children (no any single mapped to the StateNode). And a new `StateNode` is added as a child.
</blockquote></details>

Have added a test.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1029 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kjmbABILpUlUe19Bn:-L8liNZz75vYPtCVxMM4:b7vss1i) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1029)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> c

I understand that this is just an index but `c`  is very confusing. 
Either `j` or better: extract nested cycle to a separate method.
</blockquote></details>

Added `i` and `j` (in the correct order) as indexes to avoid confusion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1033 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kp2Jl2I9apen_o_XV:-L8lUvefE_cyGrpWAMZc:b4t80c6) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1033)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> if (object instanceof StateNode) {

That should always happen.
Either no check at all or assertion instead.
</blockquote></details>

Then I ask about the API design behind this. If it should always be StateNode, why the `get` method returns an Object?
In any case, added a direct cast without checking.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1035 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kpD-V7kbfD0NJhVMz:-L8lhaHQ797PsTQaNEGP:b-lwwscl) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1035)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  if (clientNode.equals(stateNode.getDomNode())) {

Do you need double navigation?

`((StateNode)mappedNodeChildren.get(0)).getNode()` will give you the very first DOM node which is mapped as a `StateNode`.

And this should return the last DOM `Node` if  there is no children at all I think.
</blockquote></details>

The `((StateNode)mappedNodeChildren.get(0)).getDomNode()` doesn't give necessarily the first node on the DOM that has a StateNode attached to it. When just running it, the `testInsertChild_recalculateIndex` and `testInsertChild` tests fail.

---

*[flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/template/ChildOrderView.java, line 64 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3773#-L8kr1xy2QMGWDMAK1sf:-L8lkrmJ6XfWb-c6xlOa:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/7f0323c4a1c62bf2eec5bb2602da84303161e205/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/template/ChildOrderView.java#L64)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>         addChildToContainer1.addClickListener(event -> {
>             Div div = new Div();
>             div.setText("Server child "
>                     + (containerWithElement.getElement().getChildCount() + 1));
>             containerWithElement.add(div);
>         });

As it shown in practice it's always good to have a test when action is done immediately in CTOR instead of as a click (or any other) listener.
It usually covers a listener case as well.
So please make at least one action immediately without a listener.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3773)*
<!-- Sent from Reviewable.io -->
